File: kotlin/src/main/kotlin/com/gildedrose/GildedRose.kt
package com.gildedrose

class GildedRose(var items: Array<Item>) {

    fun updateItems() {
        items.forEach { updateItem(it) }
    }

    private fun updateItem(item: Item) = when {
        item.name == "Aged Brie"                                 -> updateCheese(item)
        item.name == "Backstage passes to a TAFKAL80ETC concert" -> updateBackstagePasses(item)
        item.name == "Sulfuras, Hand of Ragnaros"                -> updateLegendary(item)
        else                                                     -> updateNormalItem(item)
    }

    private fun updateNormalItem(item: Item) {
        item.decreaseQuality()
        item.decreaseSellIn()
        if (item.sellIn < 0) {
            item.decreaseQuality()
        }
    }

    private fun updateCheese(item: Item) {
        item.increaseQuality()
        item.decreaseSellIn()
        if (item.sellIn < 0) {
            item.increaseQuality()
        }
    }

    @Suppress("UNUSED_PARAMETER")
    private fun updateLegendary(item: Item) {
    }

    private fun updateBackstagePasses(item: Item) {
        item.decreaseSellIn()
        if (item.sellIn < 0) {
            item.quality = 0
            return
        }

        item.increaseQuality()
        if (item.sellIn < 10) {
            item.increaseQuality()
        }
        if (item.sellIn < 5) {
            item.increaseQuality()
        }
    }

    private fun Item.increaseQuality() {
        if (this.quality < 50) {
            this.quality += 1
        }
    }

    private fun Item.decreaseQuality() {
        if (this.quality > 0) {
            this.quality -= 1
        }
    }

    private fun Item.decreaseSellIn() {
        this.sellIn -= 1
    }

}

File: csharp/GildedRose.cs
using System.Collections.Generic;
using ApprovalUtilities.Utilities;
using csharp.Strategies;

namespace csharp
{
    public class GildedRose
    {
        private const string AgedBrie = "Aged Brie";
        private const string BackstagePasses = "Backstage passes to a TAFKAL80ETC concert";
        private const string Sulfuras = "Sulfuras, Hand of Ragnaros";
        private const string Conjured = "Conjured Mana Cake";

        private readonly IList<Item> _items;

        private static readonly Dictionary<string, IItemStrategy> ItemStrategies = new Dictionary<string, IItemStrategy>
        {
            { Sulfuras, new SulfurasStrategy() },
            { AgedBrie, new AgedBrieStrategy() },
            { BackstagePasses, new BackstagePassesStrategy() },
            { Conjured, new ConjuredStrategy() }
        };

        private static readonly ItemStrategy DefaultItemStrategy = new ItemStrategy();

        public GildedRose(IList<Item> items)
        {
            _items = items;
        }

        public void UpdateQuality()
        {
            _items.ForEach(HandleItem);
        }

        private static void HandleItem(Item item)
        {
            GetItemStrategy(item).Handle(item);
        }

        private static IItemStrategy GetItemStrategy(Item item)
        {
            return ItemStrategies.ContainsKey(item.Name) ? ItemStrategies[item.Name] : DefaultItemStrategy;
        }
    }
}